### PR TITLE
[MNT] CI Modifications

### PR DESCRIPTION
**Pull request by yarnabrina**
Related #5101, #5475

* this fixes the "lower coverage" issue by adding the `test_all` workflow, which tests everyting minus the module conditional tests and always runs
* now, if anything changes in `base`, it also runs all module jobs

### CODE REVIEW

**Comment by yarnabrina** [comment]
FYI @fkiraly related to today's discussions. Let me know what you and others decide to be the final plan.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks! Looks good to me. I will add some tests for `networks`.

Can we split the changes to `_meta.py` form this PR into a second one?

**Comment by yarnabrina** [comment]
> Can we split the changes to _meta.py form this PR into a second one?

Done in #5502. Also added "ignore" flag as @benHeid suggested. Please check.

**Comment by fkiraly** [comment]
thx! Will look into this in detail after the release.

File: .github/workflows/test_all.yml
Comment (fkiraly) [inline .github/workflows/test_all.yml:24]: hm, why is this commented?

Would this solve the problem that cancel does not get run when the jobs - that would get cancelled - block the CI budget?

File: .github/workflows/test_all.yml
Comment (fkiraly) [inline .github/workflows/test_all.yml:127]: is there a more elegant way to do this?

**Review by fkiraly — COMMENTED** [review]
Nice!

Can you kindly check whether my understanding is correct:

* this fixes the "lower coverage" issue by adding the `test_all` workflow, which tests everyting minus the module conditional tests and always runs
* now, if anything changes in `base`, it also runs all module jobs
* your plan is, after merging this, we remove `test-full` from `test.yml`, yes?

I do not understand the change in `validate_datasets`, what does this do?

**Comment by fkiraly** [comment]
hm, 🤔 could we perhaps merge things so they appear as a single job in the actions register?

File: .github/workflows/test_all.yml
Comment (yarnabrina) [reply .github/workflows/test_all.yml:24]: I added it for that purpose only, but it does not seem to work. Unless the job starts immediately, i.e. if it's pending for some time becuase runners are busy, those get skipped.

I really don't understand why, asked here: https://github.com/orgs/community/discussions/73503.

Comment (fkiraly) [reply .github/workflows/test_all.yml:24]: 🤔

File: .github/workflows/test_all.yml
Comment (yarnabrina) [reply .github/workflows/test_all.yml:127]: Well, one obvious option is extra per dependency and not do this job at all. Other than that, I have no idea. May be @benHeid has some?

**Comment by yarnabrina** [comment]
>     * this fixes the "lower coverage" issue by adding the `test_all` workflow, which tests everyting minus the module conditional tests and always runs

Yes. But admittedly, I don't really like it, and will post another extra based alternative suggestion in the issue.
 
>     * now, if anything changes in `base`, it also runs all module jobs

Yes.
 
>     * your plan is, after merging this, we remove `test-full` from `test.yml`, yes?

Yes. Also, the `code-quality` job. And I'll request you, @benHeid , @achieveordie, @hazrulakmal and other core developers to evaluate the remaining `test.yml` components to see if those are really necessary to be executed for every single PR or not. If not complete removal, I'll request to consider to switch at least a subset of those to scheduled CRON jobs.

> I do not understand the change in `validate_datasets`, what does this do?

That one replaced #5004. It is dedicated to testing of `sktime/datasets`.

> hm, 🤔 could we perhaps merge things so they appear as a single job in the actions register?

I am not clear what you meant here. If it's like why not showing as a seperate workflow instead of nesting under code quality, I have an explanation what causes it from GHA, but don't know how to fix it. Asked in GHA community, let's see.

**Comment by fkiraly** [comment]
>> I do not understand the change in `validate_datasets`, what does this do?

> That one replaced #5004. It is dedicated to testing of `sktime/datasets`.

I get that, i.e., that being the rationale behind having `validate_datasets` overall.

What I do not get is what the changes to `validate_datasets` do, and what the reason is for them.

**Comment by fkiraly** [comment]
> to evaluate the remaining `test.yml` components to see if those are really necessary to be executed for every single PR or not

I would think so - otherwise we would reduce test coverage at current (true for all the remaining steps afaik).

**Comment by yarnabrina** [comment]
> What I do not get is what the changes to `validate_datasets` do, and what the reason is for them.

The changes in `on` section of `validate_datasets` (and `validate_extras`) are to meet the requirements from #5530. Instead of being triggered directly based on PR creation, they get the option to be triggered from another workflow, and that allows me to trigger it from `test_code_quality` one only if that passes.

The other changes were failed attempts for concurrency, can be ignored for now.

**Review by fkiraly — CHANGES_REQUESTED** [review]
OK, thx. Great, I understand now and this really addresses most of the oustanding points we were discusing!

For clarity and avoidance of confusion, and to make changes minimal, could you kindly remove the commented out concurrency lines?

**Comment by yarnabrina** [comment]
Concurrency changes are removed, added a check so that datasets get tested when just base changes (on logic that base change should test everything), and rebased from `main` branch.

Updated #5502 from this branch as well for diagnostics.

**Review by fkiraly — APPROVED** [review]
Thanks!

**Pull request merged**
